### PR TITLE
Fix up some colorspace logic of hierarchical color spaces

### LINKED ISSUE #605: Fix hierarchical target color space detection during shader generation
Issues:
1. currently only the root is being used as the target sometimes. If e.g. you have a root w/o colorspace and an include file with a target color space. The include file's target is ignored.
2. if there are ancestors with a target space different from the one on an element then it will not be detected since it will never traverse up the tree. It should start from the parent for the target.

### PR DESCRIPTION
**Pull request by bernardkwok**
Fix #605 

Modified logic for import and shadergen (marked as **new**)

_Imported color space logic:_
- "Transfers" target <colorspace> of the root of the imported Document.
- if import root specifies a <colorspace>:
   - Scan direct children of root:
       - if child is an <output> node, replace color space as target (**new** to fix incorrect behaviour)
       - else if child has no <colorspace> then set the <colorspace> to be that of the import root.

_Shader Generation target search logic:_
- If input element has a (source) color space, then find the target
   - If global settings has a <colorspace> use that as the target
   - Else If document has a <colorspace> use that as the target
   - Else:
       - Traverse  upwards for the highest level parent element with a <colorspace to use as a target. (**new**)
           - Fail if none found -- Without traversal will current fall into this condition for imported documents whose parent does not have a colorspace.

### CODE REVIEW

File: source/MaterialXCore/Document.cpp [hunk 1]
```diff
@@ -169,9 +170,12 @@ void Document::importLibrary(const ConstDocumentPtr& library, const CopyOptions*
         {
             childCopy->setGeomPrefix(library->getGeomPrefix());
         }
-        if (!childCopy->hasColorSpace() && library->hasColorSpace())
+        if (!libraryColorSpace.empty())
         {
-            childCopy->setColorSpace(library->getColorSpace());
+            if (!childCopy->hasColorSpace() || childCopy->isA<Output>())
```
Comment (bernardkwok) [inline source/MaterialXCore/Document.cpp]: An output's final color space should be the documents colorspace if it was specified. This information was being "lost".

File: source/MaterialXCore/Element.cpp [hunk 2]
```diff
@@ -326,6 +326,23 @@ ConstElementPtr Element::getRoot() const
...
 
+const string& Element::getActiveColorSpace(bool returnClosest) const
+{
+    ConstElementPtr colorSpaceElement = nullptr;
+    for (ConstElementPtr elem = getSelf(); elem; elem = elem->getParent())
+    {
+        if (elem->hasColorSpace())
+        {
+            colorSpaceElement = elem;
+            if (returnClosest)
```
Comment (bernardkwok) [inline source/MaterialXCore/Element.cpp]: The logic changes is to allow continue traverse to the highest level parent to find possible local targets.

File: source/MaterialXGenShader/ShaderGraph.cpp [hunk 3]
```diff
@@ -437,6 +437,14 @@ ShaderGraphPtr ShaderGraph::create(const ShaderGraph* parent, const string& name
         outputSocket->makeConnection(newNode->getOutput());
         outputSocket->setPath(shaderRef->getNamePath());
 
+        string targetColorSpace;
```
Comment (bernardkwok) [inline source/MaterialXGenShader/ShaderGraph.cpp]: Minor cleanup to not keep on doing the same evaluation for every input/param

File: source/MaterialXGenShader/ShaderGraph.cpp [hunk 4]
```diff
@@ -1107,14 +1113,26 @@ void ShaderGraph::setVariableNames(GenContext& context)
...
+        return;
+    }
+
     ShaderInput* shaderInput = shaderNode->getInput(input->getName());
     const string& sourceColorSpace = input->getActiveColorSpace();
     if (shaderInput && !sourceColorSpace.empty())
     {
         if(shaderInput->getType() == Type::COLOR3 || shaderInput->getType() == Type::COLOR4)
         {
+            // Findest ancestor closest to the root global target color space is empty
```
Comment (bernardkwok) [inline source/MaterialXGenShader/ShaderGraph.cpp]: New parent traversal check if no global colorspace is specified.

File: source/MaterialXTest/Element.cpp [hunk 5]
```diff
@@ -31,6 +31,10 @@ TEST_CASE("Element", "[element]")
     REQUIRE(elem2->getActiveColorSpace() == doc->getColorSpace());
     elem2->setColorSpace("srgb_texture");
     REQUIRE(elem2->getActiveColorSpace() == elem2->getColorSpace());
+    REQUIRE(elem2->getActiveColorSpace(false) == doc->getColorSpace());
```
Comment (bernardkwok) [inline source/MaterialXTest/Element.cpp]: Parent has no colorspace but document does so should get that back.

File: source/MaterialXTest/Element.cpp [hunk 6]
```diff
@@ -31,6 +31,10 @@ TEST_CASE("Element", "[element]")
...
+    mx::ElementPtr elem3 = elem2->addChildOfCategory("generic");
+    elem3->setColorSpace("acescg");
+    REQUIRE(elem3->getActiveColorSpace(false) == doc->getColorSpace());
```
Comment (bernardkwok) [inline source/MaterialXTest/Element.cpp]: Parent has a colorspace but the highest level parent which has a colorspace is the document so return that.

**Comment by ashwinbhat** [comment]
LGTM

**Pull request merged**